### ISSUE TITLE
Specs finish but cypress never exits

### ISSUE THREAD

**Issue by rooby** [cypress-io/cypress#20478]
### Current behavior

I am trying to get my tests running in AWS CodeBuild and the tests are all running but then it just hangs at the end.
It works as expected when I run it on my local machine though.

This is what runs after the table of test results (AWS CodeBuild - Debian 9.13):

```
2022-03-04T04:44:40.750Z cypress:webpack close /codebuild/output/src123456789/src/bitbucket.org/org/project/cypress/integration/tools/my-tool.spec.ts
2022-03-04T04:44:40.793Z cypress:server:cypress about to exit with code 2
2022-03-04T04:44:40.794Z cypress:server:browsers browsers.kill called with no active instance
2022-03-04T04:44:40.795Z cypress:proxy:http:util:prerequests metrics: lots of metrics...
2022-03-04T04:44:40.830Z cypress:cli child event fired { event: 'exit', code: 2, signal: null }
2022-03-04T04:44:40.831Z cypress:cli Stopping Xvfb
```

Then it hangs.

This differs from when it works successfully on my local machine, where it does this (local machine Windows 10 pro):
```
2022-03-04T06:13:58.653Z cypress:webpack close  C:\path\to\my\project\cypress\integration\tools\my-tool.spec.ts
2022-03-04T06:13:58.700Z cypress:server:cypress about to exit with code 2
2022-03-04T06:13:58.702Z cypress:server:browsers browsers.kill called with no active instance
2022-03-04T06:13:58.702Z cypress:proxy:http:util:prerequests metrics: lots of metrics...
2022-03-04T06:13:58.776Z cypress:cli child event fired { event: 'exit', code: 2, signal: null }
2022-03-04T06:13:58.776Z cypress:cli child event fired { event: 'close', code: 2, signal: null }
```
Then the script exits.

### Desired behavior

Exit when the tests are finished.

### Test code to reproduce

It's running like this:
```
set -e -o pipefail
cypress run --browser chrome 2>&1 | tee "cypress/logs/log.txt"
exit $?
```

### Cypress Version

9.5.1

**Comment by DannyvanHolten** [comment]
Same problem here after upgrading to Cypress 9.5.1

**Comment by BryonSullivan** [comment]
I am running into the same issue. It does seem that they will eventually finish though. I am unable to figure out what is triggering it though.

**Comment by BryonSullivan** [comment]
This issue persists in 9.5.2

**Comment by rooby** [comment]
I have now also tested 9.4.1, 9.3.1, 9.2.1 and 8.7.0 and the problem occurs in all those versions.
Within the next couple of days I will try some other systems, to see if it is specific to my OS or something.

**Comment by rooby** [comment]
This is semi-related, since it seems to be xvfb that is contributing to the problem.
https://github.com/cypress-io/cypress/issues/19868

**Comment by rooby** [comment]
Is there anything I can do to help progress this issue?
It's kind of a blocker us for using cypress and Chrome.

I see the tag has been moved to backlog, which apparently indicates current sprint, but I don't know what that means in terms of a timeframe for a fix or workaround since I don't how long your sprints are and there are 629 open backlog issues.

**Comment by tommy0831** [comment]
There is a bug introduced in Chrome where it downloads in the background: https://bugs.chromium.org/p/chromium/issues/detail?id=1319251&q=OptimizationGuidePredictionModels&can=2.
You can go to chrome://download-internals in chrome to see the downloads.

In my case, I run API tests that completes before the background downloads finishes.  When Cypress tries to quit Chrome after the run, and the download is not finished yet, Chrome will show a popup that blocks the quit, thus leaving Cypress hanging.

I tried doing cy.intercept, but it won't work because these are background requests and never get intercepted.  So, I ended up adding some wait in after().  Ideal solution is Chrome to fix the bug, a good workaround is Cypress to provide a way to force quit browser.

**Comment by rooby** [comment]
@tommy0831 Are you running your tests with headless chrome when you see that?
Are you getting the hanging on `cypress:cli Stopping Xvfb` as per the original bug report?

**Comment by tommy0831** [comment]
> @tommy0831 Are you running your tests with headless chrome when you see that? Are you getting the hanging on `cypress:cli Stopping Xvfb` as per the original bug report?

I'm running headless using cypress.run().  For me, I don't even get `cypress:cli Stopping Xvf`.  I get the below two events and it just hangs:

2022-03-04T06:13:58.776Z cypress:cli child event fired { event: 'exit', code: 2, signal: null }
2022-03-04T06:13:58.776Z cypress:cli child event fired { event: 'close', code: 2, signal: null }

**Comment by rooby** [comment]
@tommy0831 Are you running on Linux? If not you won't get the `Stopping Xvfb` line.
I suspect your issue might be a bit different to the original post of this issue, as this one hangs on Xvfb and never gets to this:

`cypress:cli child event fired { event: 'close', code: 2, signal: null }`

**Comment by tommy0831** [comment]
> @tommy0831 Are you running on Linux? If not you won't get the `Stopping Xvfb` line. I suspect your issue might be a bit different to the original post of this issue, as this one hangs on Xvfb and never gets to this:
> 
> `cypress:cli child event fired { event: 'close', code: 2, signal: null }`

I use one of the base image and install Cypress on it

FROM cypress/browsers:node16.14.0-chrome99-ff97

**Comment by rooby** [comment]
> I use one of the base image and install Cypress on it
> 
> FROM cypress/browsers:node16.14.0-chrome99-ff97
Interesting. I'll have a play around with some different images and see what results I get.
I had been planning to do that but hadn't got around to it yet.

**Comment by rooby** [comment]
@tommy0831 possibly yours is not logging the `Stopping Xvfb` line because you don't have DEBUG logging enabled. 
I was running with `export DEBUG=cypress:*`

If I follow the links to related issues in that Chrome bug you posted previously I get to this 
https://bugs.chromium.org/p/chromium/issues/detail?id=1311753
which is supposed to be fixed in Chrome v102.
I have tried now with Chrome v102.0.5005.61 and the problem persists. Not sure if that's because the Chrome fix didn't actually fix the original issue you linked to, or if it's unrelated all together.

**Comment by rooby** [comment]
I also tried Cypress 9.7.0, 8.6.0 and 8.5.0 (which was previously reported to have resolved it for someone else).
The problem persists with all versions.

**Comment by rooby** [comment]
I added xvfb to my debug logging and I got this (also added to original post)

```
2022-05-27T04:55:20.925Z cypress:cli child event fired { event: 'exit', code: 1, signal: null }
2022-05-27T04:55:20.925Z cypress:cli Stopping Xvfb
2022-05-27T04:55:20.925Z xvfb restoring process.env.DISPLAY variable
2022-05-27T04:55:20.926Z xvfb lock filename /tmp/.X99-lock
2022-05-27T04:55:20.926Z xvfb lock file /tmp/.X99-lock
2022-05-27T04:55:20.936Z xvfb lock file /tmp/.X99-lock not found when stopping
```

**Comment by ceelian** [comment]
Did anybody found a workaround so far?

**Comment by ceelian** [comment]
Okay I got it working again in the CI by pinning cypress to 9.6.1

The error I got on the Drone CI (it is an Angular/NX project):

```
nx run web-app-e2e:e2e 
[628:0604/112203.514883:ERROR:sandbox_linux.cc(377)] InitializeSandbox() called with multiple threads in process gpu-process.
[628:0604/112203.517903:ERROR:gpu_memory_buffer_support_x11.cc(44)] dri3 extension not supported.
Could not find a Cypress configuration file.

We looked but did not find a cypress.json file in this folder: /drone/src/Frontend/apps/web-app-e2e
```

And yes, the file cypress.json exists in that folder.

**Comment by rooby** [comment]
@ceelian That error "Could not find a Cypress configuration file." doesn't seem like the same error this issue is about.

**Comment by hthetiot** [comment]
@rooby are you using angular 14+ ?

I found out that reverting migration from angular 14 to angular 13 did fix the issue for me.
I'm going to attempt to provide a public repository that demonstrate that the issue is when using angular 14+ and typescript 4.7 only. I have tried many other workaround, none of them had an effect.

**Comment by rooby** [comment]
> @rooby are you using angular 14+ ?

Still on 13 so far. Thanks for the idea though, I might try playing around with the versions of a few things.

**Comment by hthetiot** [comment]
I think it has to do with this change https://github.com/angular/angular-cli/pull/23245

Here is the changes that I had to make revert migration from angular 13 to 14 to fix the cypress exit on end of spec.
This are the only change that actually have an effect.

package.json diff
```
   ..
"dependencies": {
-    "@angular/animations": "^14.0.5",
-    "@angular/common": "^14.0.5",
-    "@angular/compiler": "^14.0.5",
-    "@angular/core": "^14.0.5",
-    "@angular/forms": "^14.0.5",
-    "@angular/platform-browser": "^14.0.5",
-    "@angular/platform-browser-dynamic": "^14.0.5",
-    "@angular/router": "^14.0.5",
-    "@angular/service-worker": "^14.0.5",
-    "@angular/youtube-player": "^14.0.4",
+    "@angular/animations": "^13.3.4",
+    "@angular/common": "^13.3.4",
+    "@angular/compiler": "^13.3.4",
+    "@angular/core": "^13.3.4",
+    "@angular/forms": "^13.3.4",
+    "@angular/platform-browser": "^13.3.4",
+    "@angular/platform-browser-dynamic": "^13.3.4",
+    "@angular/router": "^13.3.4",
+    "@angular/service-worker": "^13.3.4",
+    "@angular/youtube-player": "^13.3.4",
...
"devDependencies": {
-    "@angular-devkit/build-angular": "^14.0.5",
-    "@angular-devkit/schematics": "^14.0.5",
-    "@angular-eslint/builder": "^14.0.0",
-    "@angular-eslint/eslint-plugin": "^14.0.0",
-    "@angular-eslint/eslint-plugin-template": "^14.0.0",
-    "@angular-eslint/schematics": "^14.0.0",
-    "@angular-eslint/template-parser": "^14.0.0",
-    "@angular/cli": "^14.0.5",
-    "@angular/compiler-cli": "^14.0.5",
+    "@angular-devkit/build-angular": "^13.3.3",
+    "@angular-devkit/schematics": "^13.3.3",
+    "@angular-eslint/builder": "^13.2.1",
+    "@angular-eslint/eslint-plugin": "^13.2.1",
+    "@angular-eslint/eslint-plugin-template": "^13.2.1",
+    "@angular-eslint/schematics": "^13.2.1",
+    "@angular-eslint/template-parser": "^13.2.1",
+    "@angular/cli": "^13.3.3",
+    "@angular/compiler-cli": "^13.3.4
-   "typescript": "~4.7.4",
+   "typescript": "~4.5.4",
...
```

You will need to also revert this ng14 form related changes:
```
- import { UntypedFormBuilder, UntypedFormGroup, Validators } from '@angular/forms';
+ import { FormBuilder, FormGroup, Validators } from '@angular/forms';
```

**Comment by rooby** [comment]
@hthetiot I'm already on Angular 13 (currently 13.2.x). I've not used 14 yet.
Also, that issue you linked was merged on the 8th of June and this issue was opened on the 4th of March.

Do you have a log from when you were getting the error, with this environment variable set? `DEBUG=cypress:xvfb,xvfb`
If so, does yours stop at the same part of the xvfb shutdown stage as the error in the original post of this issue?
If not, the issue you've seen might be better reported as its own issue, so it can get the attention of the cypress devs.

**Comment by jtneal** [comment]
I'm experiencing this issue after upgrading a project from Angular 12 to Angular 14. I haven't updated any of the cypress packages, so we're still using the exact same versions that worked previously, but it just hangs after the tests complete.

When run with the DEBUG info, the last things I see in the log are as follows:

```
cypress:server:cypress about to exit with code 1 +6s
cypress:server:browsers browsers.kill called with no active instance +315ms
cypress:proxy:http:util:prerequests metrics: { browserPreRequestsReceived: 13, proxyRequestsReceived: 0, immediatelyMatchedRequests: 0, eventuallyReceivedPreRequest: [], neverReceivedPreRequest: [] } +0ms
cypress:cli child event fired { event: 'exit', code: 1, signal: null } +9s
cypress:cli child event fired { event: 'close', code: 1, signal: null } +1ms
```

At this point it just hangs and never exists.

**Comment by rooby** [comment]
@jtneal your issue is different to this issue. The bug this issue relates to causes the process to hand in this step
`cypress:cli child event fired { event: 'exit', code: 1, signal: null } +9s`
and this next part never runs at all:
`cypress:cli child event fired { event: 'close', code: 1, signal: null } +1ms`
Your log also doesn't mention xvfb.

This issue is also not specifically related to Angular 14.
I suggest opening a separate issue for your issue, if there is not already one, for the best chance of a resolution to both our issues.

**Comment by h4de5** [comment]
I just want to point out, that this still seems to be a problem.

running cypress in headless mode (successfully since months). it stopped working somewhen around Dec. 22 - till now I do not have any idea what has been changed, but the issue seems to similar to the one described here:

```
  cypress:server:cypress from argv [ '/root/.cache/Cypress/10.6.0/Cypress/Cypress', '--no-sandbox', '--run-project', 'apps/dashboard-e2e', '--config', '{"baseUrl":"https://dashboard:1234/"}', '--config-file', 'cypress.config.ts', '--env', '{"tsConfig":"/tmp/npm/apps/dashboard-e2e/tsconfig.e2e.json"}', '--output-path', '/tmp/tmp-5802-kn4MqI02zz6Z', '--record', 'false', '--testing-type', 'e2e', '--cwd', '/tmp/npm', '--userNodePath', '/usr/bin/node', '--userNodeVersion', '16.18.0' ] got options { _: [ '/root/.cache/Cypress/10.6.0/Cypress/Cypress' ], sandbox: false, runProject: 'apps/dashboard-e2e', config: { e2e: { baseUrl: 'https://dashboard:1234/' }, env: { tsConfig: '/tmp/npm/apps/dashboard-e2e/tsconfig.e2e.json' } }, configFile: 'cypress.config.ts', outputPath: '/tmp/tmp-5802-kn4MqI02zz6Z', record: false, testingType: 'e2e', cwd: '/tmp/npm', userNodePath: '/usr/bin/node', userNodeVersion: '16.18.0', invokedFromCli: true, projectRoot: '/tmp/npm/apps/dashboard-e2e' } +11ms
  cypress:server:cypress scaling electron app in headless mode +0ms
  cypress:cli child event fired { event: 'exit', code: null, signal: 'SIGILL' } +1s
  cypress:cli detecting arch { osPlatform: 'linux', osArch: 'x64' } +1s
  cypress:cli arm uname -m result: { stdout: 'x86_64' }  +14ms
  cypress:cli child event fired { event: 'close', code: null, signal: 'SIGILL' } +16ms
  cypress:cli Stopping Xvfb +1s
  xvfb restoring process.env.DISPLAY variable +1s
  xvfb lock filename /tmp/.X99-lock +0ms
  xvfb lock file /tmp/.X99-lock +0ms
  xvfb lock file /tmp/.X99-lock not found when stopping +12ms
The Test Runner unexpectedly exited via a exit event with signal SIGILL
```

> ng version
```
Angular CLI: 14.2.1
Node: 16.18.0
Package Manager: npm 7.24.2 
OS: linux x64

Angular: 14.2.0
... animations, cdk, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1402.1
@angular-devkit/build-angular   14.2.1
@angular-devkit/core            14.2.1
@angular-devkit/schematics      14.2.1
@angular/cli                    14.2.1
@angular/flex-layout            14.0.0-beta.40
@schematics/angular             14.2.1
ng-packagr                      14.2.0
rxjs                            7.5.7
typescript                      4.7.4
webpack                         5.75.0
```
> npx nx --version
```
14.6.1
```

> cypress version
```
Cypress package version: 10.6.0
Cypress binary version: 10.6.0
Electron version: 19.0.8
Bundled Node version: 16.14.2
```


for some reason beyond my understanding, it does work correctly when I start the `ng` command directly:

> ng e2e dashboard-e2e --configuration production

works, tests and runs through just fine.

but if I start it prefixed with npx (or run the npm script for that very same command):

> npx ng e2e dashboard-e2e --configuration production

it results in the error above.

**Comment by hthetiot** [comment]
I migrated to @cypress/schematic@2.5.0 and it solved my issue with angular 14 and spec not ending.

TLDR:
> ng add @cypress/schematic@2.5.0

See details: https://github.com/cypress-io/cypress/tree/develop/npm/cypress-schematic

**Comment by jennifer-shehane** [comment]
Thanks for the detailed reports, everyone. The issue here was likely an orphaned browser helper process keeping the child’s stdio pipes open, which prevented the CLI’s event loop from draining (exit fired but close never did).

This is resolved in current Cypress: the CLI now resolves on the child’s exit event and calls process.exit(code) explicitly, so a lingering browser process can no longer hang the run. Closing as fixed. Please open a new issue if you still see this on a recent version.
